src/main.cpp

```diff
@@ -6975,12 +6989,18 @@ bool static ProcessMessage(const CChainParams& chainparams, CNode* pfrom, string
 
     else if (strCommand == "verack")
     {
+        LOCK(cs_main);
+        CNodeState* state = State(pfrom->GetId());
```

Comment:
Why are we certain that `state` will not be a null pointer at this point? I note that elsewhere in this method, we rely upon that invariant, but it's not obvious to me why it holds.

Reply:
The `State` function can only return null when `pnode` is not a key in `mapNodeState`. There are other cases in handling messages received from a peer `CNode *pfrom` where we assume that `State(pfrom->GetId())` will be non-null. For example, in the processing of "version":

```c++
        // Potentially mark this peer as a preferred download peer.
        {
            LOCK(cs_main);
            UpdatePreferredDownload(pfrom, State(pfrom->GetId()));
        }
```
where
```c++
void UpdatePreferredDownload(CNode* node, CNodeState* state)
{
    nPreferredDownload -= state->fPreferredDownload;
[...]
```

However, in investigating this I decided it would be a good idea to backport bitcoin/bitcoin#10176, and also to defensively add an assertion (to make the case of a null pointer here a crash, rather than undefined behaviour).